### PR TITLE
Invert parametric mask when needed on reset

### PR DESCRIPTION
**Pull request by hlecleme**
When a parametric mask channel is reset (and thus disabled), it may need to be inverted to be considered as a no-operation.

This is a proposition to improve the reset as described in https://github.com/darktable-org/darktable/pull/8248#issuecomment-781683323

As I'm not very proficient in GTK programming, @dterrahe, could you have a look to see if this is implemented in the right way? Thanks in advance!

### CODE REVIEW

**Comment by dterrahe** [comment]
Just looking at the code, without compiling, this seems to be the "right way" of implementing it in GTK, so no problems there.

I'm just wondering if it is worth the extra effort. I can see why you would argue that this is the more _correct_ way of doing this, because otherwise, if you responded to _blendop_blendif_sliders_callback when the sliders are moved to the extreme/original positions to also reset polarity, you might reset it when actually the user really wanted to disable the channel this way.

But you could argue the same thing for a user who double clicks on the slider to reset it to make sure the whole channel is _really_ disabled. And then could be surprised to find polarity reset.

My gut feeling (I could be _completely_ wrong) is that somebody who wants to disable a channel by clicking polarity doesn't mess with the actual sliders at all, either manually moving or double-click resetting them) or if they did, they would notice the polarity switching. So the simplest implementation, using existing value-changed, would possibly also be the most predictable one for them.

But maybe a stronger argument to reset polarity when the sliders are moved manually to their extreme positions is that it might help the absolute novice user, who hardly understands masks at all yet. They may have accidentally clicked on the polarity button but not noticed the state change (we used to have two buttons, input and output, so the difference might have been slightly more obvious then, but still very subtle). They probably also don't know about double click (or realise that it would do something special to polarity) so they'd naive think that moving all the sliders so that they look like in all the other channels should fix their problem. Helping them by also changing polarity for them would be helpful. The expert who knows about polarity and used it to disable the channel will surely notice this and can easily disable again.

**Comment by hlecleme** [comment]
Thanks @dterrahe for the detailed feedback, it's appreciated.

I'll try first to explain a little bit how the parametric masking is currently working, because I'm not sure to use the terms correctly and that can cause confusion.

- When parametric masking is enabled (i.e. when the parametric mask channel sliders are visible), all channels are *always* considered to compute the final mask to use for the blending operation.
- When the parametric mask channel slider markers are at their extremes (two on the left and two on the right) the selected pixels depend on the polarity of the parametric mask channel. If the polarity is positive all pixels are selected, if the polarity is negative, none of the pixels are selected.
- All the parametric mask channels are combined using the method defined in “combine mode”. If exclusive, it corresponds to the intersection of all individual parametric mask channels, if inclusive, it corresponds to the union of all individual parametric mask channels.

What usually happens when playing with the “combine mode” and the polarity of the channels, is that we could have one parametric mask channel include all pixels (as the sliders are at their extreme values) and with the inclusive “combine mode” (i.e. union) it will seem like the other channels have no effect anymore.

If the intention of a users operation is to effectively “disable” the parametric mask channel (i.e. consider it as not included in the mask computation), we then need to adapt the polarity otherwise it could select all or none of the pixels (like describe above). There is not really a “reset” value for the polarity. The default value on module reset (positive polarity, i.e. select all pixels) happens to match fine with the default “combine mode” (exclusive, i.e. intersection of channels).

So I tried to reset the slider on value change:

```diff
diff --git a/src/develop/blend_gui.c b/src/develop/blend_gui.c
index 1a7592f16..5ca0b6ee6 100644
--- a/src/develop/blend_gui.c
+++ b/src/develop/blend_gui.c
@@ -705,13 +705,26 @@ static void _blendop_blendif_sliders_callback(GtkDarktableGradientSlider *slider
     gtk_label_set_text(label[k], text);
   }
 
+  gboolean disable_channel = FALSE;
+
   /** de-activate processing of this channel if maximum span is selected */
   if(parameters[1] == 0.0f && parameters[2] == 1.0f)
+  {
     bp->blendif &= ~(1 << ch);
+    disable_channel = TRUE;
+    if(bp->mask_combine & DEVELOP_COMBINE_INCL)
+      bp->blendif |= (1 << (16 + ch));
+    else
+      bp->blendif &= ~(1 << (16 + ch));
+  }
   else
+  {
     bp->blendif |= (1 << ch);
+  }
 
   dt_dev_add_history_item(darktable.develop, data->module, TRUE);
+  if(disable_channel)
+    _blendop_blendif_update_tab(data->module, data->tab);
 }
 
 static void _blendop_blendif_polarity_callback(GtkToggleButton *togglebutton, dt_iop_gui_blend_data_t *data)
```

This reacts a bit weird for the moment: markers could be blocked at the extremes when dragging and on button release some other marker would move. This is probably due to the `_blendop_blendif_update_tab` (which does maybe too many things), but in any case the markers are changed if the polarity changes, which means that I'll probably need to adapt additional code in the widget to correctly handle the case.

I agree that for the (absolute) novice user, the masking in darktable is very confusing. The mask is composed of five to seven parametric mask channels (with polarity) that are not all visible at the same time and a set of shapes, and all those pieces are combined together using that “combine mode” parameter.

If both widgets (the slider and the polarity button) are considered as one entity (the button changes the presentation of the slider), then it could make sense to also adapt the polarity when resetting the slider. The “double click behavior to reset” behavior is already largely used throughout the darktable user interface and it is also documented in the tool-tip for the parametric mask channel slider.

Anyways, I'm interested in hearing more (user) feedback for this before spending time in fixing the weird behaviors of the alternative solution.

**Comment by dterrahe** [comment]
I see all of your points and dis-/advantages to all approaches.

Maybe a different idea could be to not change the polarity, but to always ignore 0-1 slider settings in one channel if they would overrule all other channels (either force no blending or 100% blending; if you use a mask, that is never what you want, right?). Rather than have this happen silently, the polarity icon could be made 50% opaque using `gtk_widget_set_opacity(data->filter[in_out].polarity, 0.5);` to indicate its position is irrelevant.

A more "official" way to do this would be to use `gtk_toggle_button_set_inconsistent` but that doesn't seem to have any visual effect at the moment so would probably need CSS work (I don't know what).

Longer term, maybe there should be a visual indication in the channel tab header to show which ones have an impact. At the moment, the easiest way to find out might be to look at the history tooltip. Which a newbie won't know either.

**Comment by hlecleme** [comment]
Yes, thought about that also.

Unfortunately, I don't think it is possible to change the behavior to always ignore the parametric mask channel when the markers are their extreme positions because of the compatibility with old edits. If an old edit contains a wrong polarity for one of the channels (thus generating a full or empty mask), it may be rendered differently if the channel is ignored because of this new rule.

The visual indication in the channel tab header is a good idea, I'll see what I can do for that.

**Comment by dterrahe** [comment]
> because of the compatibility with old edits

_IF_ we were convinced this was the right way to go (big IF) we could introduce a new blending parameters version and let dt_develop_blend_legacy_params either switch off blending or switch to uniform blending with opacity 0% (or some other way to achieve the same degenerated blending effect).

> I'll see what I can do for that

It would have to be a (fore or background) color or font type change because adding any marker character (*, °) would take up extra space that is already at a premium in that row. And any style change like that ideally should be themeable with CSS, which is where the problem comes in...

**Comment by hlecleme** [comment]
@TurboGit, what's your opinion on this? Should I include the suggestion of @dterrahe (cleaning the degenerated use of parametric masking in the `dt_develop_blend_legacy_params` conversion and ignore polarity when the parametric mask channel slider markers are at the extreme values) in #8217?

**Comment by TurboGit** [comment]
> If both widgets (the slider and the polarity button) are considered as one entity (the button changes the presentation of the slider), then it could make sense to also adapt the polarity when resetting the slider. The “double click behavior to reset” behavior is already largely used throughout the darktable user interface and it is also documented in the tool-tip for the parametric mask channel slider.

Yes, I fully agree with that. So the reset to change **slider + polarity** seems good to me.

> Should I include the suggestion of @dterrahe (cleaning the degenerated use of parametric masking in the `dt_develop_blend_legacy_params` conversion and ignore polarity when the parametric mask channel slider markers are at the extreme values)

At first it seems ok to me but what about the new JzCshz channels which are unbounded already and have a boost factor slider. Won't this change breaks those channels?

**Comment by hlecleme** [comment]
The parametric masking is implemented with the same algorithms in all blending “color spaces”, only the transformations to get a property of the pixels change (JzCzhz instead of HSL, etc.). Therefore, if it works with *RGB (display)* it will also work with *RGB (scene)*.

Personally, I'm not 100% confident with that cleanup, because it will be difficult to test the migration of all edge cases of previous releases, but I would be glad to try.

In the mean time, what shall we do with this PR?

**Comment by dterrahe** [comment]
I only halfheartedly suggested introducing a new version and legacy conversion _IF_ it was the only solution to all the poblems here. I don't think it is now, and I think it would be a risky move for not too much benefit. Not an advocate...

**Review by TurboGit — APPROVED** [review]
Looks good to me and works well! Thanks.

**Comment by elstoc** [comment]
This has a documentation-pending label on it but I'm really not clear on what it does. Please could someone propose a PR for dtdocs or make some suggestions to help me understand.

**Comment by hlecleme** [comment]
@elstoc I'm not sure these PRs really need to be documented. They are improvements to try to reduce the burden for some operations.

1. This PR:

   - before: when inverting the mask combine mode (between inclusive and exclusive, 3.5.3.4 typical use cases) the parametric mask channel polarities are not changed, therefore if any of the parametric mask channels is not used, the whole image will be either masked or not masked.
   - after: when inverting the mask combine mode (between inclusive and exclusive) the “unused” parametric mask channel polarities are changed to keep those masks from interfering. It is thus easier to switch between inclusive and exclusive as there is no need to go through all the parametric mask channels to fix the polarities.

2. PR #8248:
   - before: when using the color picker to set the parameters of a parametric mask channel (3.5.3.3 color pickers) some of the pixels inside the region could be excluded.
   - after: algorithms have been optimized to a) keep pixels whose values are outside of the limits and b) invert the hue channel polarity when trying to pick a magenta and orange region to avoid including green and blue pixels too.
   
   Note that this has a side effect: the color picker will now set the parameters and the polarity of the parametric mask channel to always include the pixels inside the selected region.

**Comment by elstoc** [comment]
> I'm not sure these PRs really need to be documented

Thanks for the explanations - I think you're right. I'll remove the documentation label from these PRs. We can always revisit later if we need to.
